### DISCUSSION TITLE
Seeking support setting up simple menu button

### DISCUSSION THREAD

**Discussion by sskinner-quizlet** [adobe/react-spectrum#1778] [Q&A]
I am trying to set up a simple menu button example using https://react-spectrum.adobe.com/react-aria/useMenuTrigger.html as a reference.  When I copy this example code exactly, {item.rendered} is always undefined, and no items show in the menu popup.  MenuItem is called 3 times as expected (since there are 3 Items as children--copy, cut, paste), but since {item.rendered} is undefined, nothing renders.  What am I missing here?  Is there additional setup needed to get the example working?  Thanks in advance!

**Comment by snowystinger** [comment]
Can you share a codesandbox showing the problem you're having? https://codesandbox.io/

**Reply by sskinner-quizlet** [reply]
Thanks so much for the quick response!  Interestingly, when I copy my code exactly into the sandbox it worked: https://codesandbox.io/s/sharp-wright-8pce?file=/src/App.js.   I've attached screenshots of what the [...state.collection] looks like in the working version (sandbox) and in the non working version (my local dev environment).  Any guidance on how to debug would be much appreciated, thank you!

![Screen Shot 2021-04-09 at 2 15 14 PM](https://user-images.githubusercontent.com/44449363/114241913-d70e6900-993e-11eb-9bda-250cfa923464.png)

![Screen Shot 2021-04-09 at 2 15 24 PM](https://user-images.githubusercontent.com/44449363/114241924-d970c300-993e-11eb-8de7-03fc4f1d5aeb.png)

**Reply by LFDanLu** [reply]
Forked your sandbox and it builds w/ React 16.14 now (https://codesandbox.io/s/tender-gould-9ywyu?file=/src/App.js) (no idea why yours won't build, weird). Are the versions of aria menu + button the same between your project and the sandbox? What other react-spectrum/stately/aria packages are installed locally? Might be worth digging into your node_modules and seeing what version of the react-stately packages are there since those are responsible for generating the collection

**Reply by sskinner-quizlet** [reply]
I was just starting to go down this route :).  I had installed each of the dependencies used in the example (ex: react-stately/menu, react-aria/interactions, react-stately/tree..) but then noticed that was not necessary in the sandbox.  Are the only dependencies actually needed to be installed in this example react-aria/button and react-aria/menu?  Or do we need to manually install all other dependencies as well?

side note--I did eventually get the sandbox to work with react 16.14, sorry about that confusion!

**Reply by LFDanLu** [reply]
I guess technically you only need to install react-aria/button and react-aria/menu since their dependencies bring in everything else required but I would still manually install the other packages (or just use the monopackage for [stately](https://www.npmjs.com/package/react-stately) and [aria](https://www.npmjs.com/package/react-aria))). You can always npm dedupe/yarn dedupe if need be

**Reply by sskinner-quizlet** [reply]
I just cleared my cache with `yarn cache clean`, deleted all files in node_modules, and reinstalled all the dependencies via yarn add (`"@react-aria/button": "^3.3.1",
    "@react-aria/focus": "^3.2.4",
    "@react-aria/interactions": "^3.3.4",
    "@react-aria/menu": "^3.2.0",
    "@react-aria/overlays": "^3.6.2",
    "@react-aria/utils": "^3.7.0",
    "@react-stately/collections": "^3.3.1",
    "@react-stately/menu": "^3.2.1",
    "@react-stately/tree": "^3.1.3",`), with react-aria/button and react-aria/menu being the same versions as in the sandbox.

Unfortunately still experiencing the same issue, and not sure where to go from here.  Would really appreciate any guidance on ideas of what to try next, have been pretty stuck on this and not being able to display menu items is definitely a big problem for us.  Thank you in advance!

**Reply by snowystinger** [reply]
Just to make absolutely sure, and thank you for bearing with us on this. But can you run npm ls or yarn list with appropriate args to make sure you only have one of each package installed?

Otherwise, you can start going into your node_modules and start logging things out and setting breakpoints. I'd be paying attention to places where there's context.

**Reply by sskinner-quizlet** [reply]
I checked and there were actually a few duplicates, all of them dev dependencies from @react-aria/button.  I removed react-aria/button and re-added, and those duplicates disappeared.  However, still getting the issue.

I've logged a bunch of things and noticed that getFullNode gets suspended:

![Screen Shot 2021-04-10 at 12 11 22 AM](https://user-images.githubusercontent.com/44449363/114262056-ba9c1c00-9992-11eb-9545-8b461c841fc9.png)

The code to create the full node is never reached:  https://github.com/adobe/react-spectrum/blob/887fff79e8f33c0f8953dc38955419feddd56b6a/packages/%40react-stately/collections/src/CollectionBuilder.ts#L166. 

One thing I noticed is that the recursive getFullNode in CollectionBuilder (https://github.com/adobe/react-spectrum/blob/887fff79e8f33c0f8953dc38955419feddd56b6a/packages/%40react-stately/collections/src/CollectionBuilder.ts#L129) looks different locally:

```
 let nodes = this.getFullNode(_babelRuntimeHelpersExtends({}, childNode, {
          key: nodeKey,
          index,
          wrapper: $f9fc6bc72d334a59f09d47517f9a056a$var$compose(partialNode.wrapper, childNode.wrapper)
        }), this.getChildState(state, childNode), parentKey ? "" + parentKey + element.key : element.key, parentNode);
```

Is it possible that babel is causing an issue?  Any things come to mind on where to go from here?

Appreciate your guidance, this is definitely a tricky one.

**Reply by snowystinger** [reply]
How does your project setup differ from the one you get when you download the codesandbox to run locally?
Are you using webpack? parcel? cra (also webpack but with predefined stuff)? I think given that we've now verified you don't have duplicates of any of the packages anymore, the next thing is to see where you diverge in terms of configuration.
I wouldn't bother in the code so much anymore. I was having you look there in order to ascertain if there was truly only one copy of each.
Things to now keep an eye out for would be, are you running babel on node_modules, because we've already transpiled, so you shouldn't need to.
Given that it works in the codesandbox and locally (as @LFDanLu verified) we just need to make the project configurations as close as possible until it breaks. Then you should have your answer.

**Reply by sskinner-quizlet** [reply]
Thanks so much!  Almost there--removing babel from running on react-stately in node_modules has resulted in useTreeState returning the correct thing in my environment.  The next thing that's going wrong is [...state.collection] is not returning an array of the items, but rather an array of length 1 that contains a keymap:  

<img width="903" alt="Screen Shot 2021-04-12 at 1 36 46 PM" src="https://user-images.githubusercontent.com/44449363/114458645-4d57d900-9b94-11eb-9be5-cbe99e5f8e4e.png">

Again this is only happening in my environment, not the sandbox.  Wondering if you have any ideas of what could cause a spread operator for an iterable to not return the items?  Thank you in advance!

**Reply by sskinner-quizlet** [reply]
ah--changing from `[...state.collection]` to `Array.from(state.collection)` as what is being iterated on did the trick!  

Thanks so much for your help :)

**Reply by sskinner-quizlet** [reply]
@snowystinger Is react aria transpiled as well, or is it just react stately?  Wondering if we should exlude react aria from us running babel as well.

**Reply by snowystinger** [reply]
correct, all of our packages have been
